cpp/ql/lib/experimental/semmle/code/cpp/dataflow/ProductFlow.qll

```diff
@@ -0,0 +1,119 @@
+import semmle.code.cpp.ir.dataflow.DataFlow
+import semmle.code.cpp.ir.dataflow.DataFlow2
+
+module ProductFlow {
+  abstract class Configuration extends string {
+    bindingset[this]
+    Configuration() { any() }
+
+    /**
+     * Holds if `(source1, source2)` is a relevant data flow source.
+     *
+     * `source1` and `source2` must belong to the same callable.
+     */
+    abstract predicate isSourcePair(DataFlow::Node source1, DataFlow::Node source2);
+
+    /**
+     * Holds if `(sink1, sink2)` is a relevant data flow sink.
+     *
+     * `sink1` and `sink2` must belong to the same callable.
+     */
+    abstract predicate isSinkPair(DataFlow::Node sink1, DataFlow::Node sink2);
+
+    predicate hasFlowPath(
+      DataFlow::PathNode source1, DataFlow2::PathNode source2, DataFlow::PathNode sink1,
+      DataFlow2::PathNode sink2
+    ) {
+      isSinkPair(sink1.getNode(), sink2.getNode()) and
+      reachablePair2(this, source1, source2, sink1, sink2)
+    }
+  }
+
+  private import Internal
+
+  module Internal {
+    class Conf1 extends DataFlow::Configuration {
+      Conf1() { this = "Conf1" }
+
+      override predicate isSource(DataFlow::Node source) {
+        exists(Configuration conf | conf.isSourcePair(source, _))
+      }
+
+      override predicate isSink(DataFlow::Node sink) {
+        exists(Configuration conf | conf.isSinkPair(sink, _))
+      }
+    }
+
+    class Conf2 extends DataFlow2::Configuration {
+      Conf2() { this = "Conf2" }
+
+      override predicate isSource(DataFlow::Node source) {
+        exists(Configuration conf, DataFlow::Node source1 |
+          conf.isSourcePair(source1, source) and
+          any(Conf1 c).hasFlow(source1, _)
+        )
+      }
+
+      override predicate isSink(DataFlow::Node sink) {
+        exists(Configuration conf, DataFlow::Node sink1 |
+          conf.isSinkPair(sink1, sink) and any(Conf1 c).hasFlow(_, sink1)
+        )
+      }
+    }
+  }
+
+  private predicate reachablePair1(
+    Configuration conf, DataFlow::PathNode source1, DataFlow2::PathNode source2,
+    DataFlow::PathNode node1, DataFlow2::PathNode node2
+  ) {
+    reachablePair(conf, source1, source2, node1, node2)
+    or
+    exists(DataFlow::PathNode mid1 |
+      reachablePair1(conf, source1, source2, mid1, node2) and
+      mid1.getASuccessor() = node1 and
+      mid1.getNode().getEnclosingCallable() = node1.getNode().getEnclosingCallable()
+    )
+  }
+
+  private predicate reachablePair2(
+    Configuration conf, DataFlow::PathNode source1, DataFlow2::PathNode source2,
+    DataFlow::PathNode node1, DataFlow2::PathNode node2
+  ) {
+    reachablePair1(conf, source1, source2, node1, node2) // TODO: restrict more
```

Comment:
I get what you're trying to do here - and if it works as a prototype that's great. But it looks quite problematic from a performance point of view.

Here's what I'd do instead:
Consider the set of steps from one of the two underlying flow graphs.
Join those with `.getEnclosingCallable()` for both ends and filter to only keep the inter-procedural cases. (Let's leave flow through recursive function calls as future work and ignore it for now).
Think about how well those pairs of callables can be expected to identify the step - I'd guess pretty well.
So the same for the other underlying flow.
Given our assumption about low callable-pair-to-PathNode-pair fanout in the above, we can join on the callable-pairs and produce the set of inter-procedural steps in the product flow graph.
With these we can avoid materialising relations like `reachablePair2`, which includes almost the cartesian product of co-callable-located reachable nodes, and instead only step to the relevant node-pairs.
I don't think there's a reason not to just use fastTC on the local steps, so I'd just do that.

Reply:
I've implemented this suggestion in [b1de54e](https://github.com/github/codeql/pull/9997/commits/b1de54e9aafa485579ac57fc3e4d41c16386219f)

Reply:
Great, this looks much better now. There are still some things that can be improved performance-wise, but it might be good enough for now.